vm/devices/firmware/firmware_uefi/src/service/diagnostics.rs

```diff
@@ -0,0 +1,476 @@
+    Timestamp(u64),
+    #[error("Expected message length < {0:#x}, got: {1:#x}")]
+    MessageLength(u16, u16),
+    #[error("Failed to read from buffer slice")]
+    SliceRead,
+    #[error("Encountered arithmetic overflow: {0}")]
+    Overflow(#[from] anyhow::Error),
+    #[error("Failed to read UTF-8 string: {0}")]
+    Utf8Error(#[from] std::str::Utf8Error),
+    #[error("message_end ({0:#x}) exceeds buffer slice length ({1:#x})")]
+    BadMessageEnd(usize, usize),
+}
+
+/// Represents a single parsed entry from the EFI diagnostics buffer
+struct EntryData<'a> {
+    /// The debug level of the log entry
+    debug_level: u32,
+    /// Timestamp of when the log entry was created
+    time_stamp: u64,
+    /// The boot phase that produced this log entry
+    phase: u16,
+    /// The log message itself
+    message: &'a str,
+    /// The size of the entry in bytes (including alignment)
+    entry_size: usize,
+}
+
+/// Parse a single entry from a buffer slice
+fn parse_entry(buffer_slice: &[u8]) -> Result<EntryData<'_>, EntryParseError> {
+    // Try to parse an entry from the buffer slice
+    let (entry, _) = AdvancedLoggerMessageEntryV2::read_from_prefix(buffer_slice)
+        .map_err(|_| EntryParseError::SliceRead)?;
+
+    // Validate the entry signature
+    let signature = entry.signature.to_le();
+    if signature != SIG_ENTRY {
+        return Err(EntryParseError::SignatureMismatch(SIG_ENTRY, signature));
+    }
+
+    // Validate the entry timestamp
+    if entry.time_stamp == 0 {
+        return Err(EntryParseError::Timestamp(entry.time_stamp));
+    }
+
+    // Validate the message length
+    if entry.message_len > MAX_MESSAGE_LENGTH {
+        return Err(EntryParseError::MessageLength(
+            MAX_MESSAGE_LENGTH,
+            entry.message_len,
+        ));
+    }
+
+    // Copy packed fields to local variables to avoid unaligned access
+    let message_offset = entry.message_offset;
+    let message_len = entry.message_len;
+
+    // Calculate message start and end offsets
+    let message_start = message_offset as usize;
+    let message_end = message_start
+        .checked_add(message_len as usize)
+        .ok_or_else(|| {
+            EntryParseError::Overflow(anyhow!(
+                "message_start ({}) + message_len ({})",
+                message_start,
+                message_len
+            ))
+        })?;
+
+    // Validate message end fits within the buffer slice
+    if message_end > buffer_slice.len() {
+        return Err(EntryParseError::BadMessageEnd(
+            message_end,
+            buffer_slice.len(),
+        ));
+    }
+
+    // Get the message
+    let message = std::str::from_utf8(&buffer_slice[message_start..message_end])?;
+
+    // Calculate base offset (entry header size + message length)
+    let base_offset = size_of::<AdvancedLoggerMessageEntryV2>()
+        .checked_add(message_len as usize)
+        .ok_or_else(|| {
+            EntryParseError::Overflow(anyhow!(
```

Comment:
To be clear about our error philosophy: you _either_ have an enum with all error conditions explicit, _or_ you use `anyhow`. You don't use a mix, where you stuff a string-based error into your typed error.

What would be an acceptable middle ground would be to have something like `EntryParseError::Overflow("message_len")`, where you specify the field that caused the overflow via a `&static str`. The extra noise of having the size expression in the string and the extra values is not really going to be that useful in diagnosing this condition (which let's be honest, will never happen), and it's just extra work and line noise and binary size in the meantime.